common/rust/cctrusted_base/src/tdx/report.rs

```diff
@@ -69,7 +256,174 @@ impl Tdx {
             .finalize()
             .as_slice()
             .try_into()
-            .expect("[generate_tdx_report_data] Wrong length of report data");
+            .expect("[generate_tdx_report_data] Wrong length of data");
         Ok(base64::encode(hash_array))
     }
+
+    pub fn parse_td_report(
+        report: &Vec<u8>,
+        tdx_version: TdxVersion,
+    ) -> Result<TDReport, anyhow::Error> {
+        let report_mac_struct = unsafe {
+            transmute::<[u8; 256], ReportMacStruct>(
+                report[0..256]
+                    .try_into()
+                    .expect("slice with incorrect length"),
+            )
+        };
+        let tee_tcb_info = TeeTcbInfo::new(report[256..495].to_vec(), tdx_version.clone());
+        let reserved = report[495..512].try_into().unwrap();
+        let td_info = TdInfo::new(report[512..1024].to_vec(), tdx_version.clone());
+        Ok(TDReport {
+            report_mac_struct,
+            tee_tcb_info,
+            reserved,
+            td_info,
+        })
+    }
+}
+
+#[cfg(test)]
+mod test_generate_tdx_report_data {
+    use super::*;
+    use crate::tdx::common::Tdx;
+
+    #[test]
+    //generate_tdx_report allow optional nonce
+    fn test_generate_tdx_report_data_no_nonce() {
+        let result = Tdx::generate_tdx_report_data(None, Some("YWJjZGVmZw==".to_string()));
+        assert!(result.is_ok());
+    }
+
+    #[test]
+    //generate_tdx_report allow optional data
+    fn tdx_get_quote_report_data_no_data() {
+        let result = Tdx::generate_tdx_report_data(Some("IXUKoBO1XEFBPwopN4sY".to_string()), None);
+        assert!(result.is_ok());
+    }
+
+    #[test]
+    //generate_tdx_report allow empty data
+    fn test_generate_tdx_report_data_data_size_0() {
+        let result = Tdx::generate_tdx_report_data(
+            Some("IXUKoBO1XEFBPwopN4sY".to_string()),
+            Some("".to_string()),
+        );
+        assert!(result.is_ok());
+    }
+
+    #[test]
```

Comment:
One question on the following 4 tests. Are they targeted to verify if the inputs are base64 encoded? Why two more case for short input that are not base64 encoded?

Reply:
the cases with "123" as short input is also not valid base64 encoded format, so these two cases are targeted for error scenarios where user input something short and they also missing the base64 encode the short input.

Reply:
updated to remove unnecessary user cases